tfhe/src/integer/oprf.rs

```diff
@@ -0,0 +1,187 @@
+
+        assert!(
+            1 << random_bits_count <= range_size,
+            "The range asked for a random value (=[0, 2^{random_bits_count}[) does not fit in the available range [0, {range_size}[", 
+        );
+
+        assert!(self.message_modulus().0.is_power_of_two());
+        let message_bits_count = self.message_modulus().0.ilog2() as u64;
+
+        let mut deterministic_seeder = DeterministicSeeder::<ActivatedRandomGenerator>::new(seed);
+
+        let seeds: Vec<Seed> = (0..num_blocks)
+            .map(|_| deterministic_seeder.seed())
+            .collect();
+
+        let blocks = seeds
+            .into_par_iter()
+            .enumerate()
+            .map(|(i, seed)| {
+                let i = i as u64;
+
+                if i * message_bits_count < random_bits_count {
+                    // if we generate 5 bits of noise in n blocks of 2 bits, the third (i=2) block
+                    // must have only one bit of random
+                    if random_bits_count < (i + 1) * message_bits_count {
+                        let top_message_bits_count = random_bits_count - i * message_bits_count;
+
+                        assert!(top_message_bits_count <= message_bits_count);
+
+                        self.key
+                            .generate_oblivious_pseudo_random(seed, top_message_bits_count)
+                    } else {
+                        self.key
+                            .generate_oblivious_pseudo_random(seed, message_bits_count)
+                    }
+                } else {
+                    self.key.create_trivial(0)
+                }
+            })
+            .collect::<Vec<_>>();
+
+        RadixCiphertext::from(blocks)
+    }
+}
+
+// Describes in which range a random signed integer should be generated
+#[derive(Copy, Clone)]
+pub enum SignedRandomizationSpec {
+    // taken uniformly in `[0, 2^random_bits_count[`
+    Unsigned { random_bits_count: u64 },
+    // taken uniformly in the full range [-2^(p-1), 2^(p-1)[
+    FullSigned,
+}
+
+impl ServerKey {
+    /// Generates an encrypted `num_block` blocks signed integer
+    /// using the given seed following the randomizer spec
+    /// The encryted value is oblivious to the server
+    /// It can be useful to make server random generation deterministic
+    ///
+    /// ```rust
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::integer::oprf::SignedRandomizationSpec;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    /// use tfhe::Seed;
+    ///
+    /// let size = 4;
+    ///
+    /// // Generate the client key and the server key:
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS, size);
+    ///
+    /// let random_bits_count = 3;
+    ///
+    /// let ct_res = sks.par_generate_oblivious_pseudo_random_signed_integer(
+    ///     Seed(0),
+    ///     SignedRandomizationSpec::Unsigned { random_bits_count },
```

Comment:
let's maybe use the Signed case here as it's the signed version

Reply:
Done